Support error reporting via exceptions

kumarak opened the pull request:
Fixes issue #338

lkorenc commented on include/vast/Util/Warnings.hpp:
```diff
@@ -47,16 +47,54 @@ VAST_RELAX_WARNINGS
+    struct error {
+      private:
+        std::string msg;
+        bool triggered = false;
```
I am confused about the name, because if we look in the `dtor`, it only throws `if triggered is true`. So either rename it or fix the values.

kumarak replied:
Removed `triggered` flag from the `error` class. My initial thought was to enable the exception only if `triggered` is set to true. But probably we don't need to have all these complexity and should have simple implementation of `error_stream`.

lkorenc commented on include/vast/Util/Warnings.hpp:
```diff
@@ -47,16 +47,54 @@ VAST_RELAX_WARNINGS
+        bool triggered = false;
+
+       public:
+        error(bool cond = true, std::string msg_ = std::string())
```
Why is there and option to set `triggered`? Is there a use-case for passing `false`?

kumarak replied:
removed from the changes

lkorenc commented on include/vast/Util/Warnings.hpp:
```diff
@@ -47,16 +47,54 @@ VAST_RELAX_WARNINGS
+
+       public:
+        error(bool cond = true, std::string msg_ = std::string())
+              : msg(std::move(msg_)), triggered(cond) {}
```
Argument can be just named `msg` (while I know some people like the codestyle of adding the `_` to the ctor VAST does not do it).

lkorenc commented on include/vast/Util/Warnings.hpp:
```diff
@@ -47,16 +47,54 @@ VAST_RELAX_WARNINGS
+        template <typename V>
+        error<T> operator<<(V&& s) {
+            msg += std::forward<V>(s);
+            return error<T>(triggered, msg);
```
```suggestion
            return error<T>(triggered, std::move(msg));
```

lkorenc replied:
Why not just return reference to `this`, make no copies/moves (as in deleted ctors/`=`) and then simply always throw in the dtor? Should streamline this.

lkorenc commented on include/vast/Util/Warnings.hpp:
```diff
@@ -47,16 +47,54 @@ VAST_RELAX_WARNINGS
+        error(bool cond = true, std::string msg_ = std::string())
+              : msg(std::move(msg_)), triggered(cond) {}
+        ~error() noexcept(false) {
+            if (triggered) {
```
You are never setting this to `false` (and in `operator<<` there is a new object being made) so in move ctor and move assign we should set it to `false`, otherwise that's a bug once multiple times `<<` is invoked.

xlauko commented on CMakeLists.txt:
```diff
@@ -176,6 +185,7 @@ if (ENABLE_PDLL_CONVERSIONS)
   message(STATUS "ENABLE_PDLL_CONVERSIONS")
   target_compile_definitions(vast_settings
     INTERFACE
+      -fexceptions
```
add this only in case of `ENABLE_VAST_EXCEPTION` is `ON`

move it to previous `target_compile_definitions`

kumarak replied:
Oops, my mistake. I was intended to add under `ENABLE_VAST_EXCEPTION` but added under different flag. Will fix this.

xlauko commented on include/vast/Util/Warnings.hpp:
```diff
@@ -47,16 +47,54 @@ VAST_RELAX_WARNINGS
 #include <llvm/Support/Debug.h>
 VAST_UNRELAX_WARNINGS
 
+#include <stdexcept>
```
move under `ENABLE_VAST_EXCEPTION`

xlauko commented on include/vast/Util/Warnings.hpp:
```diff
@@ -47,16 +47,54 @@ VAST_RELAX_WARNINGS
 {
+#ifdef ENABLE_VAST_EXCEPTION
+    template <typename T = std::runtime_error>
+    struct error {
```
rename to `error_stream`

xlauko replied:
also something like:
```
struct error_stream {
    std::stringstream buff;
    
    virtual ~error_stream() noexcept(false) {
        throw std::runtime_error(buff.str());
    }

    template< typename token_type >
    error_stream& operator<<(token_type &&token) {
        buff << std::forward< token_type >(token);
        return *this;
    }
};
```

should work?

kumarak commented:
@lkorenc, @xlauko, I addressed the comments and simplified the implementation of `error_stream` whose only job is to throw the exception in `dtor`. Having template and trigger flag was probably making it look complex.
